src/LuaSerializer.cpp

```diff
@@ -328,6 +328,261 @@ const char *LuaSerializer::unpickle(lua_State *l, const char *pos)
 	return pos;
 }
 
+void LuaSerializer::pickle_json(lua_State *l, int to_serialize, Json::Value &out, std::string key)
+{
+	LUA_DEBUG_START(l);
+
+	// tables are pickled recursively, so we can run out of Lua stack space if we're not careful
+	// start by ensuring we have enough (this grows the stack if necessary)
+	// (20 is somewhat arbitrary)
+	if (!lua_checkstack(l, 20))
+		luaL_error(l, "The Lua stack couldn't be extended (out of memory?)");
+
+	to_serialize = lua_absindex(l, to_serialize);
+	int idx = to_serialize;
+
+	if (lua_getmetatable(l, idx)) {
+		lua_getfield(l, -1, "class");
+		if (lua_isnil(l, -1))
+			lua_pop(l, 2);
+
+		else {
+			const char *cl = lua_tostring(l, -1);
+
+			out["class"] = cl;
+
+			lua_getfield(l, LUA_REGISTRYINDEX, "PiSerializerClasses");
+
+			lua_getfield(l, -1, cl);
+			if (lua_isnil(l, -1))
+				luaL_error(l, "No Serialize method found for class '%s'\n", cl);
+
+			lua_getfield(l, -1, "Serialize");
+			if (lua_isnil(l, -1))
+				luaL_error(l, "No Serialize method found for class '%s'\n", cl);
+
+			lua_pushvalue(l, idx);
+			pi_lua_protected_call(l, 1, 1);
+
+			idx = lua_gettop(l);
+
+			if (lua_isnil(l, idx)) {
+				lua_pop(l, 5);
+				LUA_DEBUG_END(l, 0);
+				return;
+			}
+		}
+	}
+
+	switch (lua_type(l, idx)) {
+		case LUA_TNIL:
+			break;
+
+		case LUA_TNUMBER: {
+			out = Json::Value(lua_tonumber(l, idx));
```

Comment:
Probably need to change this to something that doesn't lose precision. Except if that something generates a JSON string then there will also need to be a way to differentiate in the JSON output between a Lua string and a Lua number... which is a bit of a PITA.